tls/s2n_connection.c

```diff
@@ -1392,3 +1397,49 @@ int s2n_connection_set_keyshare_by_name_for_testing(struct s2n_connection *conn,
 
     POSIX_BAIL(S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
 }
+
+int s2n_connection_get_peer_cert_chain(const struct s2n_connection *conn, struct s2n_cert_chain_and_key *cert_chain_and_key)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(cert_chain_and_key);
+
+    const struct s2n_x509_validator *validator = &conn->x509_validator;
+    POSIX_ENSURE_REF(validator);
+    POSIX_ENSURE(validator->state == VALIDATED, S2N_ERR_CERT_NOT_VALIDATED);
+
+    struct s2n_cert_chain *cert_chain = cert_chain_and_key->cert_chain;
+    POSIX_ENSURE_REF(cert_chain);
+    struct s2n_cert **insert = &cert_chain->head;
+    POSIX_ENSURE(*insert == NULL, S2N_ERR_INVALID_CERT_CHAIN);
+
+    /* X509_STORE_CTX_get1_chain() returns a validated cert chain if a previous call to X509_verify_cert() was successful.
+     * X509_STORE_CTX_get0_chain() is a better API because it doesn't return a copy. But it's not available for Openssl 1.0.2.
+     * See the comments here:
+     * https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get1_chain.html
+     */
+    STACK_OF(X509) *cert_chain_validated = X509_STORE_CTX_get1_chain(validator->store_ctx);
```

Comment:
This stack will leak on any error

Reply:
Yes, I was worried about the same. I outlined solution for it in the PR description call-out section. We cannot directly use the `DEFER_CLEANUP` for this use-case as the `_free` method here takes two arguments and the signature varies to what DEFER_CLEANUP expects `k_X509_pop_free(cert_chain_validated, X509_free);`. We would have to write a static wrapper free function that could be used with `DEFER_CLEANUP` or use goto cleanup (which is not recommended!) :)